### PR TITLE
Empty log queue on flush and close

### PR DESCRIPTION
**Pull request by zeripath**
It is possible for log events to remain in the buffer off the multichannelledlog
and thus not be logged despite close or flush.

This PR simply adds a function to empty the queue before closing or flushing.
(Except when the logger is paused.)

Reference #19982

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by delvh — APPROVED** [review]

**Comment by zeripath** [comment]
> IIRC, there is also a `ChannelledLog`, which uses similar logic.

Good catch - I'd forgotten that was the structure.

File: modules/log/event.go [hunk 1]
```diff
@@ -345,18 +361,41 @@ func (m *MultiChannelledLog) Start() {
...
+func (m *MultiChannelledLog) emptyQueue() bool {
+	for {
+		select {
+		case event, ok := <-m.queue:
+			if !ok {
+				return false
+			}
+			m.rwmutex.RLock()
+			for _, logger := range m.loggers {
+				err := logger.LogEvent(event)
```
Comment (delvh) [inline modules/log/event.go:387]: Why does this one throw an error and the other doesn't?

Comment (zeripath) [reply modules/log/event.go:387]: That's the API that was there originally.

**Review by wxiaoguang — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work.

**Pull request merged**